Modules/gcmodule.c

```diff
@@ -1191,80 +1251,260 @@ handle_resurrected_objects(PyGC_Head *unreachable, PyGC_Head* still_unreachable,
+}
+
+static void
+gc_collect_young(PyThreadState *tstate,
+                 struct gc_collection_stats *stats)
+{
+    GCState *gcstate = &tstate->interp->gc;
+    PyGC_Head *young = &gcstate->young.head;
+    PyGC_Head *aging = &gcstate->old[gcstate->aging_space].head;
 #ifdef Py_STATS
-    if (_Py_stats) {
-        _Py_stats->object_stats.object_visits = 0;
+    {
+        Py_ssize_t count = 0;
+        PyGC_Head *gc;
+        for (gc = GC_NEXT(young); gc != young; gc = GC_NEXT(gc)) {
+            count++;
+        }
+        GC_STAT_ADD(0, objects_queued, count);
     }
 #endif
-    int i;
-    Py_ssize_t m = 0; /* # objects collected */
-    Py_ssize_t n = 0; /* # unreachable objects that couldn't be collected */
-    PyGC_Head *young; /* the generation we are examining */
-    PyGC_Head *old; /* next older generation */
-    PyGC_Head unreachable; /* non-problematic unreachable trash */
-    PyGC_Head finalizers;  /* objects with, & reachable from, __del__ */
-    PyGC_Head *gc;
-    _PyTime_t t1 = 0;   /* initialize to prevent a compiler warning */
-    GCState *gcstate = &tstate->interp->gc;
+    PyGC_Head survivors;
+    gc_list_init(&survivors);
+    gc_collect_region(tstate, young, &survivors, UNTRACK_TUPLES, stats);
+    Py_ssize_t surivivor_count = 0;
+    if (gcstate->aging_space) {
+        /* objects in aging space have bit set, so we set it here */
+        surivivor_count = gc_list_set_space(&survivors, 1);
+    }
+    else {
+        PyGC_Head *gc;
+        for (gc = GC_NEXT(&survivors); gc != &survivors; gc = GC_NEXT(gc)) {
+#ifdef GC_DEBUG
+            assert(gc_old_space(gc) == 0);
+#endif
+            surivivor_count++;
+        }
+    }
+    gc_list_merge(&survivors, aging);
+    validate_old(gcstate);
+    gcstate->young.count = 0;
+    gcstate->old[gcstate->aging_space].count++;
+    gcstate->incremental_gc_progress -= surivivor_count;
+    add_stats(gcstate, 0, stats);
+}
 
-    // gc_collect_main() must not be called before _PyGC_Init
-    // or after _PyGC_Fini()
-    assert(gcstate->garbage != NULL);
-    assert(!_PyErr_Occurred(tstate));
+static inline int
+IS_IN_OLDEST(PyGC_Head *gc, uintptr_t aging_space)
+{
+    assert(aging_space == 0 || flip_old_space(aging_space) == 0);
+    return gc_old_space(gc) != aging_space;
+}
+
+struct container_and_flag {
+    PyGC_Head *container;
+    uintptr_t aging_space;
+};
 
-    if (gcstate->debug & DEBUG_STATS) {
-        PySys_WriteStderr("gc: collecting generation %d...\n", generation);
-        show_stats_each_generations(gcstate);
-        t1 = _PyTime_GetPerfCounter();
+/* A traversal callback for adding to container) */
+static int
+visit_add_to_container(PyObject *op, void *arg)
+{
+    OBJECT_STAT_INC(object_visits);
+    struct container_and_flag *cf = (struct container_and_flag *)arg;
+    uintptr_t aging = cf->aging_space;
+    if (_PyObject_IS_GC(op)) {
+        PyGC_Head *gc = AS_GC(op);
+        if (_PyObject_GC_IS_TRACKED(op) &&
```

Comment:
I think including objects from the other space can lead to some problematic behavior.  If you have a large object graph which is referenced from smaller more frequently allocated objects this will continuously pull that large object graph in and blow the budget.  This is basically what I was concerned with regarding `sys.modules` - you can have a module which imports sys, you can be creating lambda's in that module which are short lived but become cyclical trash, and to collect the lambda you need to traverse the world.

I also think given this behavior the need for two different lists of objects isn't really necessary, you could instead just move the objects to the end of the collecting space and we'd get back to them when we can, and I think the behavior would be identical to the existing algorithm (except maybe we'd pick up some extra objects when we'd flip the spaces).

I think another potential problem with this behavior is that you're not eagerly adding objects in the current space to this list transitively.  That means if we visit a large object graph and blow the budget then we may not get to other transitively referenced objects that are in the space we're collecting from added to the container, and therefore despite collecting a huge object graph we still won't have collected enough to clear the cycle.

Below's a program that seems to grow unbounded, I had briefly experimented with using the `_PyGC_PREV_MASK_COLLECTING` flag here to mark objects we want to include instead of using the aging space, but that also didn't work (I would have expected on some collections we get a bunch of these little cycles, and then after a flip we need to walk the large object graph once), so I'm not certain what exactly needs to be done to fix this.

```
class LinkedList:
    def __init__(self, next=None, prev=None):
        self.next = next
        if next is not None:
            next.prev = self
        self.prev = prev
        if prev is not None:
            prev.next = self


def make_ll(depth):
    head = LinkedList()
    for i in range(depth):
        head = LinkedList(head, head.prev)
    return head


head = make_ll(10000)

while True:
    newhead = make_ll(200)
    newhead.surprise = head
```

Reply:
Any cycle present in the old generation at the start of the full scan (when we flip spaces) *will* be collected during that scan (before we flip spaces again) See https://github.com/python/cpython/pull/108038#issuecomment-1713976117 

There will always be object graphs that perform badly, but all cycles will be collected eventually (assume the program runs for long enough).


This program looks like most cycles created will be handled in the young collection, so I don't see a problem there, but if we increase `depth` so that the cycles will outlive the young collection, then it might take a while to collect the cycles, and the first increment will likely have to visit all the nodes reachable from the global variable `head`.

Reply:
I guess I wasn't clear enough (and maybe you didn't run it) because this program actually runs out of memory :)  You're right that most cycles should be collected in young, but I'm guessing one survives every young collection, and those build up and are uncollectible.  I think we probably only successfully collect 1 of the non-long lived objects every collection because we repeatedly re-process the long-lived linked list. 

If I modify the program to hold onto 100 of these at a time before clearing them out it runs out of memory even faster.

Reply:
I'm seeing the memory use grow slowly, although seemingly without bounds. So something is wrong.

There are two spaces in the old generation. `aging` and `oldest`.
After a young or incremental collection we add survivors to the end of `aging`.
We only collect cycles within the `oldest` space. After a flip, *all* objects will be in the `oldest` space, so if there are any cycles they will be collected, not moved back to the aging space.

I modified your program to include `gc.set_threshold(2000, 2, 0)` which makes the incremental collector process objects five times as fast, in which case the memory appears to stays bounded.

I was hoping to merge this and then play with thresholds, but it looks like we will need some sort of adaptive threshold before merging.

Reply:
You only collect cycles in the `oldest` space but the reason I placed this comment here is that you do gather the transitive closure from the `aging` space. Therefore think the statement "We only collect cycles within the oldest space." is incorrect given this code - once you've included a single object from `aging` you will consider its transitive closure as well.  

But including these objects seems like it should be unnecessary though, once you flip, you'll re-consider those objects and their transitive closure.  

And as I said before I think this basically eliminates any usefulness of the two spaces...  you may as well just be moving the objects to the end of the `oldest` space if you're willing to suck them into the transitive closure.

Reply:
Why do you think we collect objects from the aging space?
That is not the intention, and I don't see where that happens.